features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,557 @@
+                            unsigned char *output )
+{
+    unsigned char c;
+    unsigned char ov[17];
+    int ret = 0;
+
+    if( ctx == NULL || input == NULL || output == NULL || iv == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    while( length-- )
+    {
+        memcpy( ov, iv, 16 );
+        if ( (ret = mbedtls_aes_crypt_ecb( ctx, MBEDTLS_AES_ENCRYPT, iv, iv ) ) != 0 ) {
+            return ret;
+        }
+
+        if( mode == MBEDTLS_AES_DECRYPT )
+            ov[16] = *input;
+
+        c = *output++ = (unsigned char)( iv[0] ^ *input++ );
+
+        if( mode == MBEDTLS_AES_ENCRYPT )
+            ov[16] = c;
+
+        memcpy( iv, ov + 1, 16 );
+    }
+
+    return ret;
+}
+#endif /*MBEDTLS_CIPHER_MODE_CFB */
+
+#if defined(MBEDTLS_CIPHER_MODE_CTR)
+/*
+ * AES-CTR buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+                           size_t length,
+                           size_t *nc_off,
+                           unsigned char nonce_counter[16],
+                           unsigned char stream_block[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    size_t n = nc_off ? *nc_off : 0;
+    size_t processed = 0;
+    int ret = 0;
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || input == NULL || output == NULL || nonce_counter == NULL || stream_block == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    while ( processed < length ) {
+        if ( n > 0 ) {
+            /* start by filling up the IV */
+            output[processed] = (unsigned char)( input[processed] ^ stream_block[n] );
+            n = ( n + 1 ) & 0x0F;
+            processed++;
+            continue;
+        } else {
+            /* process one ore more blocks of data */
+            CRYPTO_TypeDef *device = crypto_management_acquire();
+            device->WAC = 0;
+            device->CTRL = CRYPTO_CTRL_INCWIDTH_INCWIDTH4;
+
+            CORE_ENTER_CRITICAL();
+            CRYPTO_KeyBufWrite(device, (uint32_t*)ctx->key, (ctx->keybits == 128UL) ? cryptoKey128Bits : cryptoKey256Bits);
+            CRYPTO_DataWriteUnaligned(&device->DATA1, (uint8_t *)nonce_counter);
+            CORE_EXIT_CRITICAL();
+
+            /* strategy: encrypt nonce, encNonce xor input -> output, inc(nonce) */
+            size_t iterations = (length - processed) / 16;
+            for (size_t i = 0; i < iterations; i++ ) {
+                device->CMD = CRYPTO_CMD_INSTR_DATA1TODATA0;
+                device->CMD = CRYPTO_CMD_INSTR_AESENC;
+                while ((device->STATUS & CRYPTO_STATUS_INSTRRUNNING) != 0);
```

Comment:
This might be a future feature best implemented using the mbedTLS threading primitives (if the api would support semaphores in addition to mutexes, since the release would happen from IRQ).

For this specific case, I doubt it would make a whole lot of sense since an AES block operation only takes about 70 clock cycles anyways. Context-switching away and back again would be more expensive. Halting the core might be an alternative.

For efficiency when using accelerators, how about adding an API for multi-block ECB operations? Setting up and tearing down the accelerator for each block is expensive.